### ISSUE TITLE
[BUG] Simple auto indent no longer works

### ISSUE THREAD

**Issue by mltony** [notepad-plus-plus/notepad-plus-plus#15396]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description of the Issue

What is simple auto indent? When I press enter i want the indentation of the next line to be exactly the same as current line. I don't want any fancy algorithms decide for me that after if statement the next line should have higher indentation level. This was the default behavior in v8.1 and I liked it. I just tried upgrading to v8.6.8 and in python files it appears to have switched to smart auto indent that I hate, and there appears to be no setting to switch back to simple auto indent (or did I miss that setting somewhere?)

To be crystal clear, here are steps to reproduce below. Also I am creating this as a bug (hope that's ok) since this is a regression compared to v8.1 and this is so critical for me it forces me to downgrade for the time being.

### Steps To Reproduce

1. Open any python file in Notepad++ v8.6.8.
2. On an empty line type:
```
def func():
    if True:
```

Then press enter.

### Current Behavior

I denote `^` to be system cursor. By switching auto indent option I can achieve either of two options:

* Option 1: By default (auto indent enabled):

```
def func():
    if True:
        ^
```

* Option 2: auto indent disabled:

```
def func():
    if True:
^
```

### Expected Behavior

I use `^` character to denote system cursor. I want it to be here:

```
def func():
    if True:
    ^
```

### Debug Information

```shell
Notepad++ v8.6.8   (64-bit)
Build time : Jun  4 2024 - 00:30:00
Path : C:\Program Files\Notepad++\notepad++.exe
Command Line : "C:\Program Files\Notepad++\change.log" 
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
Periodic Backup : ON
OS Name : Windows 11 Pro (64-bit)
OS Version : 23H2
OS Build : 22631.3296
Current ANSI codepage : 1252
Plugins : 
    mimeTools (3.1)
    NppConverter (4.6)
    NppExport (0.4)
```

### Anything else?

P.S. Thanks to the maintainers for keeping notepad++ alive and well!

**Comment by xomx** [comment]
Not a bug but a currently implemented feature: https://github.com/notepad-plus-plus/notepad-plus-plus/pull/15159

Currently, you will get the desired behavior, e.g. when the Language will be set to "None (plain text)" but not to the "Python".

**Comment by alankilborn** [comment]
Auto-indent is by design a "smart" auto indent.
The fact that Python didn't have this type of auto indenting previously was an oversight bug -- it should have had it from the start.

**Comment by mltony** [comment]
@xomx,
How to make notepad++ consider all `*.py` files to be plain text?

Also I disagree that this is a desired behavior. A quick search shows that I'm not the only one looking for simple auto indent: https://community.notepad-plus-plus.org/topic/19050/manual-indent-v-auto-indent-v-smart-indent
It might be true that most people indeed prefer smart autoindent, but it is surprising to hear that simple auto indent doesn't even deserve to be considered by the maintainers, especially given the fact that simple autoindent has been working beautifully for decades.

Most text editors out there support simple autoindent. Examples: VSCode, sublime, UltraEdit - there's probably many more.

My last argument is I am actually blind, so trying to figure out indentation level is kind of hard when you can't see and when smart autoindent adjusts it for you. This slows me down a great deal. Also this new behavior undoes decades of muscle memory.

**Comment by mltony** [comment]
I just donated $50 to notepad++. Please consider implementing a setting for simple auto indent. You can treat this as a feature request if you don't agree that this is a regression.

**Comment by alankilborn** [comment]
>My last argument is I am actually blind

You should always fully state the WHY of your desire up front (I re-read your original posting and you did not do this).  It helps others understand your need better.  Usually arguments like "it used to be like..." or "I just want..." don't motivate anyone to really consider your request.

>this new behavior undoes decades of muscle memory

Also not a good reason; sometimes muscles need retraining.

>...trying to figure out indentation level is kind of hard when you can't see and when smart autoindent adjusts it for you.

It is rather intuitive: If the line above ends with a colon (`:`), then the next line is indented.  I'd think that if you gave it a try for a little while, you'd get used to it and actually appreciate it.  This is the way it is with a lot of changes/features added to software.

>I just donated $50 to notepad++

Now we feel bribed.  :-)

Before the bribe, I had decided that if the author would accept it, I would agree to work on adding the following into the software:

```txt
(*) No auto-indent
( ) Simple auto-indent (follows Previous Line)
( ) Smart auto-indent
```

or 

```txt
Auto-indent:
(*) None
( ) Previous Line
( ) Smart

"Simple" and "Smart" would be equivalent for languages that don't support smart-indenting, which is most languages.

**Comment by xomx** [comment]
@alankilborn 
I was considering exactly the same thing (None, Simple (prev.line), Smart (or Simple)), radios or a combobox  :-)

IDK if it could be really useful for more users (?)
It seems to me as a very specific need. Especially with the Python, where the right indentation is not only for better code readability but is directly the part of the lang-syntax. (from that POV I can also see the previous smart-indent patch as a bug-fix rather than an enhancement)

**Comment by ozone10** [comment]
IMHO smart indent for python is desired behavior.
The provided example from community forum is for c family languages. And unlike python, they have more indentation styles, because they are curly-bracket and not structured languages.
Because n++ does not support all indentation styles, option to use simple block indentation makes here more sense.

@alankilborn 
Maybe something like this?
![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/55940305/4ea3238d-2c7e-42b1-a78c-7608efc1abd3)

**Comment by mltony** [comment]
yes, these triple radio buttons or combo box to select the type of auto indent would be amazing.

> Usually arguments like "it used to be like..." or "I just want..." don't motivate anyone to really consider your request.

I am struggling to come up with any better arguments. I don't want to flaunt my blindness, so using this as a last resort argument. But I also have a couple of sighted friends who also prefer simple auto indent.
Out of curiosity, what kind of better argument do you have in mind that would motivate you? Surely one can't use philosophical arguments, like it is the righteous thing to do or it is morally wrong not to implement a feature. So it always comes down to personal preferences. The thing is that I am a contributor to NVDA - an open-source screenreader project as well, and in NVDA world we judge feature requests by the number of users who show interest in them and I haven't seen any more profound arguments being made there except for who will find what feature useful.

> sometimes muscles need retraining

If you are making a change that is universally agreed that this change is an improvement, then perhaps so. Otherwise it's a matter of taste. 

> I'd think that if you gave it a try for a little while, you'd get used to it and actually appreciate it.

I tried it for about a day and still I don't like it. In fact the first editor that tried to make me use smart autoindent was Microsoft Visual Studio 6.0 in the late 1990s, and I have been fighting smart indent since then. Mostly successfullly - most editors have a way to disable smart indent.

> It is rather intuitive

Not intuitive for me. I often work with indented plain text files, so I'd need to keep simple autoindent behavior in mind. Then when I switch to a Python file, there will be another behavior. One thing is the line after `:` gets indented. Then what about statements like `return` or `break` - some smart editors unindent after those lines - does notepad++ do the same thing? Then I occasionally write some C++ code, where some other set of rules applies. As a result, I'd probably have to analyze current indentation level every time after pressing Enter key. I would rather spent resources of my brain on actual coding then on trying to predict smart autoindent logic. And it's not too hard for me to press tab where it needs to be pressed.

But I am not trying to convince you that my way is best. I am perfectly aware that we're the minority. I just thought it should be easy enough for notepadd++ to keep this minority happy.

**Comment by alankilborn** [comment]
>Out of curiosity, what kind of better argument do you have in mind that would motivate you?

The blindness argument really would have made sense to say initially; it wouldn't have been "flaunted" -- it certainly adds a different way of a sighted user understanding the difficulty with a feature that has a large component of visuality to it.

> Then what about statements like `return` or `break` - some smart editors unindent after those lines - does notepad++ do the same thing?

The feature is called "auto indent", not "auto indent and unindent".  :-)
Also, perhaps these examples are flawed as there is often code like:  `if foo: return` in which case there is clearly no unindent potential.

---

If @donho approves of @ozone10 's idea presented in his screenshot, he can assign me and I'll work on implementation.

**Comment by alankilborn** [comment]
@donho Sorry for the second ping on this, but I fear that you missed the first one above.  :-)

**Comment by donho** [comment]
@mltony 
> I just donated $50 to notepad++. Please consider implementing a setting for simple auto indent. You can treat this as a feature request if you don't agree that this is a regression.

Could you confirm me if the transaction of the donation is 29G923385T916183X please?

**Comment by mltony** [comment]
Here is what I see in my paypal receipt:
Transaction ID 4395-0823-3184-6397

**Comment by donho** [comment]
@mltony 
Under which second name please?

**Comment by mltony** [comment]
> Donation from Anton.malykh@gmail.com

My name should appear as iether Anton Malykh or Tony Malykh

**Comment by mltony** [comment]
Also I just forwarded you paypal receipt to don.h@free.fr

**Comment by donho** [comment]
@mltony 
Thank you for your confirmation.

I have processed a refund for your donation. 

![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/90293/89a3d3c6-2677-4a81-8415-d426e8901480)


As mentioned in my refund message, donations do not influence feature requests. Decisions regarding feature implementation are based on project needs, not monetary contributions. In the thread, you can observe people discussing and contributing their ideas, including code. This collaborative process informs decisions and determines which features are included in Notepad++.

If the desired feature is implemented as expected, I recommend considering a donation at a later time. Additionally, you can always donate directly to the person who implements the feature via GitHub.

Thank you regardless for your donation.

**Comment by donho** [comment]
@ozone10 
> Maybe something like this?
![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/55940305/4ea3238d-2c7e-42b1-a78c-7608efc1abd3)

Looks good. But will **"Use default value"** apply also to the group **"Auto-Indent"**. If it will, the option "Smart" is not for all languages - what's the behaviour when **"Auto-Indent"** option is **"Smart"** in **"Use default value"** but a language has no smart indent implemented?.

**Comment by ozone10** [comment]
@donho 
that is only suggestion for @alankilborn

Currently auto-indent is global, so first PR could implement only "global" option. Per language could be in next PR, if requested by users.

I would also maybe add [?] button with tooltip, which explains more about auto-indent: that smart is currently only for c-like languages and python, for other block is used instead if smart is selected.

**Comment by donho** [comment]
@ozone10 
Your suggestion is prettier graphically, but lead the confusion IMO.
I finally chose to make a new "Indentation" page in Preferences dialog and regroup 2 groups of settings together (See #15414).

@alankilborn 
> If @donho approves of @ozone10 's idea presented in his screenshot, he can assign me and I'll work on implementation.

Sorry for not assigning this feature request to you.
As I mentioned above, I think mixing both groups will make users confused. But I didn't have any idea about it.
I have had to modify several times RC file to get the result that I satisfy (#15414). If you were assigned to this FR, it would be time consuming. Anyway the PR is there. Thank you for pinging me.
